src/group/libp2p_group_relcast_server.erl

```diff
@@ -404,9 +426,33 @@ handle_inbound_message(Index, MsgKey, Msg, State=#state{handler=Handler, handler
             {Action, delete_message(MsgKey, State#state{handler_state=NewHandlerState})};
         {NewHandlerState, {send, Messages}=_Action} ->
             save_state(State, Handler, HandlerState, NewHandlerState),
-            {ok, send_messages(Messages, delete_message(MsgKey, State#state{handler_state=NewHandlerState}))}
+            {ok, send_messages(Messages, delete_message(MsgKey, State#state{handler_state=NewHandlerState}))};
+        {NewHandlerState, {close, Timeout}} ->
+            save_state(State, Handler, HandlerState, NewHandlerState),
+            erlang:send_after(Timeout, self(), force_close),
+            {ok, close_workers(State#state{close_state=closing, handler_state=NewHandlerState})}
     end.
 
+-spec close_workers(#state{}) -> #state{}.
+close_workers(State=#state{workers=Workers}) ->
+    lists:foreach(fun(#worker{pid=self}) ->
+                          ok;
+                     (#worker{pid=Pid}) ->
+                          libp2p_group_worker:close(Pid)
+                  end, Workers),
+    State.
+
+-spec close_check(#state{}) -> #state{}.
+close_check(State=#state{close_state=closing}) ->
+    case count_messages(?OUTBOUND, all, State) of
```

Comment:
Given we know that no more inbound messages will be added to the queue, can we simply count the remaining messages and decrement a counter, rather than doing an expensive count every time?

Reply:
Yeah.. the count is not a traversal over bitcask though, just a fold over the out-keys. You concerned that would take too long?

Reply:
Oh, that's probably fine then.